palace/models/waveportoperator.cpp

```diff
@@ -580,14 +580,46 @@ WavePortData::WavePortData(const config::WavePortData &data,
     }
   }
 
-  // Create vector for initial space for eigenvalue solves (for nullspace of [Aₜₜ  0]
-  //                                                                         [0   0]).
-  GetInitialSpace(*port_nd_fespace, *port_h1_fespace, port_nd_dbc_tdof_list,
-                  port_h1_dbc_tdof_list, v0);
-  e0.SetSize(v0.Size());
-  e0t.SetSize(port_nd_fespace->GetTrueVSize());
+  // Create vector for initial space for eigenvalue solves.
+  GetInitialSpace(*port_nd_fespace, *port_h1_fespace, port_dbc_tdof_list, v0);
+  e0.SetSize(port_nd_fespace->GetTrueVSize() + port_h1_fespace->GetTrueVSize());
   e0n.SetSize(port_h1_fespace->GetTrueVSize());
 
+  // The operators for the generalized eigenvalue problem are:
+  //                [Aₜₜ   Aₜₙ]  [eₜ]  = -kₙ² [Bₜₜ  0] [eₜ]
+  //                [Aₜₙᵀ  Aₙₙ] [eₙ]         [0  0] [eₙ]
+  // for the wave port of the given index. The transformed variables are related to the true
+  // field by Eₜ = eₜ and Eₙ = eₙ / ikₙ. We will actually solve the shift-and-inverse
+  // problem (A - σ B)⁻¹ B e = λ e, with λ = 1 / (-kₙ² - σ).
+  // Reference: Vardapetyan and Demkowicz, Full-wave analysis of dielectric waveguides at a
+  //            given frequency, Math. Comput. (2003).
+  // See also: Halla and Monk, On the analysis of waveguide modes in an electromagnetic
+  //           transmission line, arXiv:2302.11994 (2023).
+  mu_eps_min = 0.0;  // Use standard inverse transformation to avoid conditioning issues
+                     // associated with shift
+  // double c_max = mat_op.GetLightSpeedMax().Max();
+  // MFEM_VERIFY( c_max > 0.0 && c_max < mfem::infinity(),
+  //             "Invalid material speed of light detected in WavePortOperator!");
+  // mu_eps_min = 1.0 / (c_max * c_max);
+  std::tie(Atnr, Atni) = GetAtn(mat_op, *port_nd_fespace, *port_h1_fespace);
+  std::tie(Antr, Anti) = GetAnt(mat_op, *port_h1_fespace, *port_nd_fespace);
+  std::tie(Annr, Anni) = GetAnn(mat_op, *port_h1_fespace, port_normal);
+  {
+    // The HypreParMatrix constructor from a SparseMatrix on each process does not copy
+    // the SparseMatrix data, but that's OK since this Dnn is copied in the block system
+    // matrix construction.
+    mfem::Vector d(port_h1_fespace->GetTrueVSize());
+    d.UseDevice(false);  // SparseMatrix constructor uses Vector on host
+    d = 0.0;
+    mfem::SparseMatrix diag(d);
+    auto Dnn = std::make_unique<mfem::HypreParMatrix>(
+        port_h1_fespace->GetComm(), port_h1_fespace->Get().GlobalTrueVSize(),
+        port_h1_fespace->Get().GetTrueDofOffsets(), &diag);
```

Comment:
That's right. `HypreParMatrixFromBlocks` does not allow for a fully empty row or column so that it can generate the right block sizes from the input. So, we just explicitly pass a zero diagonal block for `Bnn`.